### DISCUSSION TITLE
Python new grammar PEG and special operator

### DISCUSSION THREAD

**Discussion by aranega** [textX/textX#412] [Q&A]
Hi @igordejanovic 

I'm trying to adapt the new [PEG grammar of Python](https://docs.python.org/3/reference/grammar.html) to make it work with textX. I'm unsure it will work at the end, but anyway, it's worth the try. There is some easy adaptation to do, but I don't know how to deal with an operator that is available in the Pytho grammar: `~`. It is said that 

```
# ~
#   Commit to the current alternative, even if it fails to parse.
#
```

And in the [related PEP](https://peps.python.org/pep-0617/#e-5): 

```
Commit to the current alternative, even if it fails to parse.

rule_name: '(' ~ some_rule ')' | some_alt
In this example, if a left parenthesis is parsed, then the other alternative won’t be considered, even if some_rule or ‘)’ fail to be parsed.
```

What would be the best way to translate this operator in textX? I guess I can rewrite the rule, but I was wondering if there was another trick that could do it?

One example of the rules is the following:

```
assignment:
    | NAME ':' expression ['=' annotated_rhs ] 
    | ('(' single_target ')' 
         | single_subscript_attribute_target) ':' expression ['=' annotated_rhs ] 
    | (star_targets '=' )+ (yield_expr | star_expressions) !'=' [TYPE_COMMENT] 
    | single_target augassign ~ (yield_expr | star_expressions) 
```

I rewrite it this way:

```
assignment:
    NAME ':' expression ('=' annotated_rhs)?
    | ('(' single_target ')'
    | single_subscript_attribute_target) ':' expression ('=' annotated_rhs)?
    | (star_targets '=' )+ (yield_expr | star_expressions) !'=' TYPE_COMMENT?
    | single_target augassign  (yield_expr | star_expressions)?
;
```
I used an optional for the rule after the the `~`, but I don't think it's the right semantic as if `yield_expr` or `start_expressions` cannot be parsed, an error will be triggered by textX.

**Comment by igordejanovic** [comment]
Hi Vincent,

If I understand `~` rule correctly what is specified before the rule is some kind of prefix which if parsed will prevent any subsequent alternative to be matched.

What you could try is something along this line (haven't tried this so might not work):

```
MyRule: Prefix Some Other Rules | !Prefix something else | !Prefix another one;
Prefix: ...;
```

Thus, if `Prefix` matches then only the first alternative would ever be considered as the others have negative lookahead assertion with the same prefix so they can't match.

Not sure what `~` does in the example above from the Python grammar as it is used in the last alternative so there is no other alternative to try.

**Comment by igordejanovic** [comment]
Moving this issue to discussions.

**Comment by aranega** [comment]
I guess your suggestion is the right one. I'm really unsure what the operator is really doing, I have some trouble to understand it properly, but your explanation and way or rewriting the rule makes a lot of sense. If that the right semantic, it means that `~` (and `[ expr ]`) is really more kind of syntactic sugar than anything else. 
I'll try to see if I can write a small grammar for the syntax they have, and write a transformation towards a pure textX grammar. By hand it's kind of long and error prone. I'll try to experiment with all of that and see where I can go from there :)
